cumulus/test/service/src/lib.rs

```diff
@@ -0,0 +1,627 @@
+	)?;
+
+	let params = PartialComponents {
+		backend,
+		client,
+		import_queue,
+		keystore_container,
+		task_manager,
+		transaction_pool,
+		select_chain: (),
+		other: (),
+	};
+
+	Ok(params)
+}
+
+/// Start a node with the given parachain `Configuration` and relay chain `Configuration`.
+///
+/// This is the actual implementation that is abstract over the executor and the runtime api.
+#[sc_tracing::logging::prefix_logs_with(parachain_config.network.node_name.as_str())]
+async fn start_node_impl<RB>(
+	parachain_config: Configuration,
+	_collator_key: Option<CollatorPair>,
+	relay_chain_config: Configuration,
+	wrap_announce_block: Option<Box<dyn FnOnce(AnnounceBlockFn) -> AnnounceBlockFn>>,
+	rpc_ext_builder: RB,
+) -> sc_service::error::Result<(
+	TaskManager,
+	Arc<Client>,
+	Arc<NetworkService<Block, H256>>,
+	RpcHandlers,
+)>
+where
+	RB: Fn(Arc<Client>) -> Result<jsonrpc_core::IoHandler<sc_rpc::Metadata>, sc_service::Error>
+		+ Send
+		+ 'static,
+{
+	if matches!(parachain_config.role, Role::Light) {
+		return Err("Light client not supported!".into())
+	}
+
+	let mut parachain_config = prepare_node_config(parachain_config);
+
+	let params = new_partial(&mut parachain_config)?;
+
+	let validator = parachain_config.role.is_authority();
+	let prometheus_registry = parachain_config.prometheus_registry().cloned();
+	let transaction_pool = params.transaction_pool.clone();
+	let mut task_manager = params.task_manager;
+
+	let primary_chain_full_node = {
+		let span = tracing::info_span!(
+			sc_tracing::logging::PREFIX_LOG_SPAN,
+			name = relay_chain_config.network.node_name.as_str()
+		);
+		let _enter = span.enter();
+
+		subspace_service::new_full::<
+			subspace_test_runtime::RuntimeApi,
+			subspace_test_client::TestExecutorDispatch,
+		>(relay_chain_config, false)
+		.await
+		.map_err(|_| sc_service::Error::Other("Failed to build a full subspace node".into()))?
+	};
+
+	let client = params.client.clone();
+	let backend = params.backend.clone();
+
+	let import_queue = cirrus_client_service::SharedImportQueue::new(params.import_queue);
+	let (network, system_rpc_tx, start_network) =
+		sc_service::build_network(sc_service::BuildNetworkParams {
+			config: &parachain_config,
+			client: client.clone(),
+			transaction_pool: transaction_pool.clone(),
+			spawn_handle: task_manager.spawn_handle(),
+			import_queue: import_queue.clone(),
+			block_announce_validator_builder: None,
+			warp_sync: None,
+		})?;
+
+	let rpc_extensions_builder = {
+		let client = client.clone();
+
+		Box::new(move |_, _| rpc_ext_builder(client.clone()))
+	};
+
+	let rpc_handlers = sc_service::spawn_tasks(sc_service::SpawnTasksParams {
+		rpc_extensions_builder,
+		client: client.clone(),
+		transaction_pool: transaction_pool.clone(),
+		task_manager: &mut task_manager,
+		config: parachain_config,
+		keystore: params.keystore_container.sync_keystore(),
+		backend: backend.clone(),
+		network: network.clone(),
+		system_rpc_tx,
+		telemetry: None,
+	})?;
```

Comment:
Maybe we don't need RPC in this test framework?

Reply:
It's necessary if you want to send some extrinsic. Although we're not using it right now, there is no need to spend extra time removing it right now and bringing it back in the future since it does not bring in much maintainance cost and does not slow down the tests obviously IMO.

Reply:
Hm, I didn't expect RPC necessary for sending extrinsics, we should have more direct way of doing it instead of going through serialization/deserialization and the whole networking stack.

Reply:
That's the way they used in polkadot&cumulus, I'm not saying it's the best way, but it can save you some time to not develop your own version if it already meets the requirement.

Reply:
Let's keep it for now, but let's also not take what Polkadot does for granted, we need to consider what makes most sense for us. If we can name things better, make tests run faster or bring less dependencies in - we should strongly consider those options.

Reply:
Yeah, I understand and totally agree with that, we definitely should consider what's the best way in our position, but realistically I have to say we don't always have that much time and resources to carry out and make it actually happen :D